### PR TITLE
Changed behaviour for layer-shift keys

### PR DESCRIPTION
**Pull request by gedankenexperimenter**
With this change, layer shift keys set `highestLayer`, but don't affect `layerState`. This means that when holding a `ShiftToLayer(N)` key, layer N becomes the top active layer, and layers above N are ignored while that key is held. This should result in less surprising behaviour for users who are unfamiliar with the layer stack.

In addition, this means that only one layer shift can be active at a time. It also means that locked layers are not deactivated when a matching layer shift key is released.

To make this change, I turned `handleKeymapKeyswitchEvent()` into a private static member function so it could access the `highestLayer()` variable. It was already hidden from other translation units, so this shouldn't be a big deal. I also added `row` & `col` as parameters so that the key in question could be masked if it detects a change to `highestLayer` (e.g. from another layer shift key being pressed -- we don't want them fighting for control).

`updateActiveLayers()` was changed so that `highestLayer` always gets a lookup, regardless of its bit in `layerState`.

### CODE REVIEW

**Comment by gedankenexperimenter** [comment]
I've only done a few quick tests on this so far, and I'm pretty sure another change will be needed to make the layer shift override a subsequently-pressed `LockLayer()` key (as long as it's still held).

File: src/layers.cpp [hunk 1]
```diff
@@ -105,15 +111,15 @@ Layer_::updateActiveLayers(void) {
...
+        if (isOn(layer) || layer == highestLayer) {
           Key mappedKey = (*getKey)(layer, row, col);
 
           if (mappedKey != Key_Transparent) {
             activeLayers[row][col] = layer;
             break;
           }
         }
-        layer--;
+        --layer;
```
Comment (gedankenexperimenter) [inline src/layers.cpp]: Oops. I changed this decrement to a prefix operator with the thought that it would avoid an unnecessary copy, but of course, the compiler already optimized that out. It shouldn't really be part of this change, but it's also harmless.

**Comment by gedankenexperimenter** [comment]
This change is now +10 bytes of program memory, +1 byte of data, by the way.

**Comment by algernon** [comment]
> In addition, this means that only one layer shift can be active at a time. It also means that locked layers are not deactivated when a matching layer shift key is released.

This is one thing I'm a bit unhappy about. I fully understand that the current layer system is confusing, because it is. But it is also very, very flexible, and allows for some [clever use cases](https://community.keyboard.io/t/layer-switching-confusion/1035/12?u=algernon), which the changed behaviour would not support.

It would be great if we could either support both cases, or lifted layering into a plugin (with the less confusing one being shipped with the factory firmware), but that's likely a bigger undertaking.

Other than that, I think the proposed behaviour is on the right track. Thank you for spending so much time and effort to think it through, describe the behaviour, and implementing it! <3

**Comment by gedankenexperimenter** [comment]
Does your clever use case involve keys on _two different shifted layers_ being visible at once? If not, this change wouldn't interfere with it.

**Comment by gedankenexperimenter** [comment]
…or it might, because of the masked key. I'm not sure how MagicCombo works.

**Comment by algernon** [comment]
`MagicCombo` checks key positions by looking at the scanner's state, so as long as at least one key is unmasked, it should still work. The interesting part of my setup is that both palms together get me to an empty layer. But your suggestion on the forums (of using MagicCombo for that too) should cover that case as well.

**Comment by algernon** [comment]
Correction: masking doesn't interfere at all, because `MagicCombo` works in the loop hook.

**Comment by algernon** [comment]
Almost a year later (sorry!) I'm still not sold on this idea completely. I still think that the current layer behaviour has its uses, and while the same could be achieved with the help of MagicCombos, that doesn't feel quite right. We'd be removing a flexible thing and replacing it with something one would have to work around to get old behaviour back.

What if we added a new set of layer shift keys, called `MoveToLayer(n)`? or brought back the old `MomentaryLayer(n)` naming? These could work the way described here, and leave the existing stuff available too. We'd transition the factory firmware to `MomentaryLayer(n)`, and voila!

**Comment by gedankenexperimenter** [comment]
In my opinion, adding another type of layer shift would add more confusion than it's worth, especially if the one with this PR's behaviour is called `MoveToLayer()` (sounds permanent) or `MomentaryLayer()` (sounds more like the existing behaviour).

I may not be recalling the details of this PR correctly, but it seems like the only thing that you wouldn't be able to do with this behaviour is temporarily turn on a lower layer, changing only the keys that aren't transparent on the top layer (or other layers in between). Do you really have a use case for that?

**Comment by gedankenexperimenter** [comment]
There was a [case on the forum](https://community.keyboard.io/t/set-auto-start-layer-for-keyboard/2547/2) in the past few days where this layer-shift behaviour was what the user expected (more or less).

**Comment by obra** [comment]
I'm closing this out. It's been hanging around forever, but we don't intend to merge it. I'm sorry it took us so long to say that definitively.
